// ra-core/dist/routing/adapters/reactRouterProvider.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.reactRouterProvider = void 0;
var React = __importStar(require("react"));
var react_1 = require("react");
var react_router_dom_1 = require("react-router-dom");
var routerProviderFuture = { v7_startTransition: false, v7_relativeSplatPath: false };
/**
 * Hook to check if navigation blocking is supported.
 * In react-router, blocking requires a data router.
 */
var useCanBlock = function () {
    var dataRouterContext = (0, react_1.useContext)(react_router_dom_1.UNSAFE_DataRouterContext);
    var dataRouterStateContext = (0, react_1.useContext)(react_router_dom_1.UNSAFE_DataRouterStateContext);
    return !!(dataRouterContext && dataRouterStateContext);
};
/**
 * Wrapper around react-router's useNavigate that returns a stable function reference.
 *
 * react-router's useNavigate forces rerenders on every navigation, even if we don't use the result.
 * @see https://github.com/remix-run/react-router/issues/7634
 *
 * This wrapper uses a ref to return a stable function reference, avoiding unnecessary rerenders
 * in components that use navigate but don't need to rerender on navigation.
 */
var useNavigate = function () {
    var navigate = (0, react_router_dom_1.useNavigate)();
    var navigateRef = (0, react_1.useRef)(navigate);
    (0, react_1.useEffect)(function () {
        navigateRef.current = navigate;
    }, [navigate]);
    // Return a stable function that always calls the latest navigate
    return React.useCallback(function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return navigateRef.current.apply(navigateRef, args);
    }, []);
};
/**
 * Internal router component that creates a HashRouter.
 * Only used when not already inside a router context.
 */
var InternalRouter = function (_a) {
    var children = _a.children, basename = _a.basename;
    var router = (0, react_router_dom_1.createHashRouter)([{ path: '*', element: React.createElement(React.Fragment, null, children) }], {
        basename: basename,
        future: {
            v7_fetcherPersist: false,
            v7_normalizeFormMethod: false,
            v7_partialHydration: false,
            v7_relativeSplatPath: false,
            v7_skipActionErrorRevalidation: false,
        },
    });
    return (React.createElement(react_router_dom_1.RouterProvider, { router: router, future: routerProviderFuture }));
};
/**
 * RouterWrapper component for react-router.
 * Creates a HashRouter if not already inside a router context.
 */
var RouterWrapper = function (_a) {
    var basename = _a.basename, children = _a.children;
    var isInRouter = (0, react_router_dom_1.useInRouterContext)();
    if (isInRouter) {
        return React.createElement(React.Fragment, null, children);
    }
    return React.createElement(InternalRouter, { basename: basename }, children);
};
/**
 * Default router provider using react-router-dom.
 * This provider is used by default when no custom routerProvider is provided to <Admin>.
 */
exports.reactRouterProvider = {
    // Hooks
    useNavigate: useNavigate,
    useLocation: react_router_dom_1.useLocation,
    useParams: react_router_dom_1.useParams,
    useBlocker: react_router_dom_1.useBlocker,
    useMatch: react_router_dom_1.useMatch,
    useInRouterContext: react_router_dom_1.useInRouterContext,
    useCanBlock: useCanBlock,
    // Components
    Link: react_router_dom_1.Link,
    Navigate: react_router_dom_1.Navigate,
    Route: react_router_dom_1.Route,
    Routes: react_router_dom_1.Routes,
    Outlet: react_router_dom_1.Outlet,
    // Router creation
    RouterWrapper: RouterWrapper,
    // Utilities
    matchPath: react_router_dom_1.matchPath,
};
